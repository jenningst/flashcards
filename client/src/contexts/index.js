import React from 'react';
import { AuthProvider } from './auth-context';
import { UserProvider } from './user-context';
import { ApolloProvider } from 'react-apollo';

import { ApolloClient, HttpLink } from 'apollo-boost';
import { InMemoryCache } from 'apollo-cache-inmemory';

const client = new ApolloClient({
  link: new HttpLink({ uri: "http://localhost:4000/graphql" }),
  cache: new InMemoryCache(),
});

function AppProviders({ children }) {
  return (
    <AuthProvider>
      <UserProvider>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </UserProvider>
    </AuthProvider>
  );
};

export default AppProviders;