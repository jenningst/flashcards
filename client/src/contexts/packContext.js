import React, { createContext, useContext, useReducer } from 'react';
import packReducer, { initialState } from '../reducers/packReducer';

const PackStateContext = createContext();
const PackDispatchContext = createContext();

function PackProvider({ children }) {
  const [state, dispatch] = useReducer(packReducer, initialState);
  return (
    <PackStateContext.Provider value={state}>
      <PackDispatchContext.Provider value={dispatch}>
        {children}
      </PackDispatchContext.Provider>
    </PackStateContext.Provider>
  );
};

function usePackState() {
  const context = useContext(PackStateContext);
  if (context === undefined) {
    throw new Error('usePackState must be used within a PackProvider')
  }
  return context;
};

function usePackDispatch() {
  const context = useContext(PackDispatchContext);
  if (context === undefined) {
    throw new Error('usePackDispatch must be used within a PackProvider')
  }
  return context;
};

export { PackProvider, usePackState, usePackDispatch };