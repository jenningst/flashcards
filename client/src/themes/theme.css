/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/