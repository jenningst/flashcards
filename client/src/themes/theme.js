export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};