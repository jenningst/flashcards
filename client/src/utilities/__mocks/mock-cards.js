export const mockCards = [
  { id: '', text: 'What do you call crystal clear urine', answer: '1080pee.' },
  { id: '', text: 'What was Forrest Gump’s email password', answer: '1forrest1.' },
  { id: '', text: 'What does a house wear', answer: 'A dress.' },
  { id: '', text: 'What do you call a bear with no teeth', answer: 'A gummy bear.' },
  { id: '', text: 'What do you call a young musician', answer: 'A minor.' },
  { id: '', text: 'What do you call a line of rabbits marching backwards', answer: 'A receding hairline.' },
  { id: '', text: 'What do you call a belt with a watch on it', answer: 'A waist of time.' },
  { id: '', text: 'What do you call a fake noodle', answer: 'An Impasta.' },
  { id: '', text: 'How much does a hipster weigh', answer: 'An instagram.' },
  { id: '', text: 'Why is Kylo Ren so angry', answer: 'Beause he’s always Ben Solo.' },
  { id: '', text: 'Why does Superman gets invited to dinners', answer: 'Because he is a Supperhero.' },
  { id: '', text: 'Why didn’t the lion win the race', answer: 'Because he was racing a cheetah.' },
  { id: '', text: 'Why did the cookie cry', answer: 'Because his father was a wafer so long!' },
  { id: '', text: 'Why did the tomato turn red', answer: 'Because it saw the salad dressing.' },
  { id: '', text: 'Why did the octopus beat the shark in a fight', answer: 'Because it was well armed.' },
  { id: '', text: 'Why can’t you take inventory in Afghanistan', answer: 'Because of the tally ban.' },
  { id: '', text: 'Why are there fences on graveyards', answer: 'Because people are dying to get in.' },
  { id: '', text: 'Why did one banana spy on the other', answer: 'Because she was appealing.' },
  { id: '', text: 'Why don’t some couples go to the gym', answer: 'Because some relationships don’t work out.' },
  { id: '', text: 'Why isn’t suntanning an Olympic sport', answer: 'Because the best you can ever get is bronze.' },
  { id: '', text: 'Why don’t cannibals eat clowns', answer: 'Because they taste funny.' },
  { id: '', text: 'Is your refrigerator running', answer: 'Better go catch it.' },
  { id: '', text: 'What did the Buffalo say to his little boy when he dropped him off at school', answer: 'Bison.' },
  { id: '', text: 'Why does the man want to buy nine rackets', answer: 'Cause tennis too many.' },
  { id: '', text: 'What do prisoners use to call each other', answer: 'Cell phones.' },
  { id: '', text: 'What do cows tell each other at bedtime', answer: 'Dairy tales.' },
  { id: '', text: 'What’s it called when you have too many aliens', answer: 'Extraterrestrials.' },
  { id: '', text: 'What do you call a beautiful pumpkin', answer: 'GOURDgeous.' },
  { id: '', text: 'What do you call a cow with no legs', answer: 'Ground beef.' },
  { id: '', text: 'What do you call Samsung\'s security guards', answer: 'Guardians of the Galaxy.' },
  { id: '', text: 'Why was dumbo sad', answer: 'He felt irrelephant.' },
  { id: '', text: 'Why did the lion eat the tightrope walker', answer: 'He wanted a well-balanced meal!' },
  { id: '', text: 'Did you hear about the guy who lost the left side of his body', answer: 'He\'s alright now.' },
  { id: '', text: 'How does Moses make coffee', answer: 'Hebrews it.' },
  { id: '', text: 'What do you call a cow with all of its legs', answer: 'High steaks.' },
  { id: '', text: 'How does a penguin build it’s house', answer: 'Igloos it together.' },
  { id: '', text: 'What do you call a girl with one leg that\'s shorter than the other', answer: 'Ilene.' },
  { id: '', text: 'Where do you imprison a skeleton', answer: 'In a rib cage.' },
  { id: '', text: 'What happens to nitrogen when the sun comes up', answer: 'It becomes daytrogen.' },
  { id: '', text: 'Did you hear about the circus that caught on fire', answer: 'It was in tents.' },
  { id: '', text: 'Did you hear about the invention of the white board', answer: 'It was remarkable.' },
  { id: '', text: 'Did you hear about the crime in the parking garage', answer: 'It was wrong on so many levels.' },
  { id: '', text: 'Have you ever tried to eat a clock', answer: 'It\'s very time consuming.' },
  { id: '', text: 'So what if I don’t know what apocalypse means', answer: 'It’s not the end of the world!' },
  { id: '', text: 'What do you mean June is over', answer: 'Julying.' },
  { id: '', text: 'What does Superman have in his drink', answer: 'Just ice.' },
  { id: '', text: 'Why couldn’t the dead car drive into the cluttered garage', answer: 'Lack of vroom.' },
  { id: '', text: 'What do you call a cow with two legs', answer: 'Lean beef.' },
  { id: '', text: 'What do you receive when you ask a lemon for help', answer: 'Lemonaid.' },
  { id: '', text: 'I got hit in the head with a can of soda', answer: 'Luckily it was a soft drink.' },
  { id: '', text: 'What’s america’s favorite soda', answer: 'Mini soda.' },
  { id: '', text: 'What cheese can never be yours', answer: 'Nacho cheese.' },
  { id: '', text: 'Did you hear about the pun that was actually funny', answer: 'Neither have we.' },
  { id: '', text: 'Want to hear a joke about paper', answer: 'Nevermind it’s tearable.' },
  { id: '', text: 'Want to hear a pizza joke', answer: 'Nevermind', answer: 'it’s too cheesy.' },
  { id: '', text: 'Can February March', answer: 'No but April May.' },
  { id: '', text: 'What do you call a french pig', answer: 'Porque.' },
  { id: '', text: 'What’s it called when you put a cow in an elevator', answer: 'Raising the steaks.' },
  { id: '', text: 'What does a dog say when he sits down on a piece of sandpaper', answer: 'Ruff!' },
  { id: '', text: 'What planet is like a circus', answer: 'Saturn it has three rings!' },
  { id: '', text: 'What happened when the magician got mad', answer: 'She pulled her hare out!' },
  { id: '', text: 'Why can\'t bicycles stand up on their own', answer: 'Since they are 2 tired.' },
  { id: '', text: 'What kind of shoes do ninjas wear', answer: 'Sneakers.' },
  { id: '', text: 'What should you do if you are cold', answer: 'Stand in the corner. It’s 90 degrees.' },
  { id: '', text: 'What did the alien say to the pitcher of water', answer: 'Take me to your liter.' },
  { id: '', text: 'How many tickles does it take to make an octopus laugh', answer: 'Ten tickles.' },
  { id: '', text: 'Want to hear a pun about ghosts', answer: 'That\'s the spirit!' },
  { id: '', text: 'Do you know where you can get chicken broth in bulk', answer: 'The stock market.' },
  { id: '', text: 'What kind of car does a sheep drive', answer: 'Their SuBAHHru.' },
  { id: '', text: 'Why do trees have so many friends', answer: 'They branch out.' },
  { id: '', text: 'You know why I like egg puns', answer: 'They crack me up!' },
  { id: '', text: 'Why don’t vampires go to barbecues', answer: 'They don’t like steak.' },
  { id: '', text: 'Why are frogs so happy', answer: 'They eat whatever bugs them.' },
  { id: '', text: 'Why should you never trust a train', answer: 'They have loco motives.' },
  { id: '', text: 'How do trees access the internet', answer: 'They log on.' },
  { id: '', text: 'Why shouldn’t you trust atoms', answer: 'They make up everything.' },
  { id: '', text: 'Have you ever heard of a music group called Cellophane', answer: 'They mostly wrap.' },
  { id: '', text: 'How do mountains see', answer: 'They peak.' },
  { id: '', text: 'Did you hear about the human cannonball', answer: 'Too bad he got fired!' },
  { id: '', text: 'Have you ever tried to milk a cow which has been cut in half', answer: 'Udder madness.' },
  { id: '', text: 'When does a farmer dance', answer: 'When he drops the beet.' },
  { id: '', text: 'Why is metal and a microwave a match made in heaven', answer: 'When they met sparks flew.' },
  { id: '', text: 'How was Rome split in two', answer: 'With a pair of Ceasars.' },
  { id: '', text: 'What do you do with a dead chemist', answer: 'You barium.' },
  { id: '', text: 'Why is a skeleton a bad liar', answer: 'You can see right through it.' },
  { id: '', text: 'What do you do to an open wardrobe', answer: 'You closet.' },
  { id: '', text: 'What happens when you eat too many spaghettiOs', answer: 'You have a vowel movement.' },
  { id: '', text: 'What do you do when balloons are hurt', answer: 'You helium.' },
  { id: '', text: 'How do you organize an outer space party', answer: 'You planet.' },
  { id: '', text: 'Do you know sign language," You should learn it', answer: 'it’s pretty handy.' },
  { id: '', text: 'What did daddy spider say to baby spider', answer: 'You spend too much time on the web.' },
];