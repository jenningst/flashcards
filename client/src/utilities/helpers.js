export function zeroPad(n, places) {
  return n.toString().padStart(3, '0');
};