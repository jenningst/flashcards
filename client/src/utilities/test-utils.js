const uuidv4 = require('uuid/v4');

function seedMockPack(pack, n = 1) {
  const filter = uuidv4(); // generate a filter
  const name = uuidv4(); // generate a name
  const modes = ['REVIEW_MODE', 'WRITE_MODE']; // TODO: remove hard-coding
  const mode = modes[getRandomInt(0, 1)]; // generate a random mode
  let numToSeed = getRandomInt(1, n || pack.length - 1); // generate a random number of cards
  
  let cards = [];
  for (let i = 0; i < numToSeed; i++) {
    let card = getRandomElement(pack, pack.length - 1);
    card.id = uuidv4();
    cards.push(card);
  }
 
  return { mode, name, filter, cards };
};

function getRandomElement(arr, ceiling) {
  return arr[getRandomInt(0, ceiling)];
};

function getRandomInt(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
};

export { seedMockPack };