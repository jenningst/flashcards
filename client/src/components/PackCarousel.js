import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { usePackDispatch } from '../contexts/pack-context';
import { useKeyPress } from '../hooks/use-key-press';
import { Mutation } from  'react-apollo';
import { GET_FLASHCARDS_BY_PACK, CREATE_FLASHCARD } from '../queries';

import { Subhead, Caption3 } from './elements/Text';
import { MediumButton } from './elements/Button';
import { ReactComponent as Back } from '../components/icons/svg/back.svg';
import Flashcard from './Flashcard';
import ComposeFlashcard from './ComposeFlashcard';

function PackCarousel({ mode, filter, cards }) {
  const dispatch = usePackDispatch();

  const [index, setIndex] = useState(0);
  const [questionText, setQuestionText] = useState('');
  const [questionAnswer, setQuestionAnswer] = useState('');
  const handleTextChange = e => setQuestionText(e.target.value);
  const handleAnswerChange = e => setQuestionAnswer(e.target.value);

  const exitToPackHome = () => dispatch({ type: 'RESET_MODE' });
  const backPress = useKeyPress(37);
  const forwardPress = useKeyPress(39);
  const spacebarPress = useKeyPress(32)

  useEffect(() => {
    if (backPress && index > 0) {
      setIndex(index - 1);
    }
  }, [backPress]);

  useEffect(() => {
    if (forwardPress && index < cards.length - 1) {
      setIndex(index + 1);
    }
  }, [forwardPress]);

  // console.log(`cards: ${cards.length}`);
  // console.log(`index: ${index}`);

  // const priorCard = () => {
  //   if (index > 0) { // question is not the first
  //     if (backPress) { // useKeyPress hook returns true
  //       setIndex(index - 1);
  //     }
  //   }
  // };

  // const nextCard = () => {
  //   if (index < cards.length) { // question is not the last
  //     if (backPress) { // useKeyPress hook returns true
  //       setIndex(index + 1);
  //     }
  //   }
  // };

  const saveCardAndRefresh = () => {
    setQuestionText('');
    setQuestionAnswer('');
    setIndex(index + 1);
  };

  const currentQuestion = cards[index];
  const isReviewMode = mode === 'REVIEW_MODE';
  
  const CreateFlashcard = () => {
    return (
      <Mutation
        mutation={CREATE_FLASHCARD}
        update={(cache, { data }) => {
          // get our current cards from the cache
          const { fetchFlashcardsByPack } = cache.readQuery({ 
            query: GET_FLASHCARDS_BY_PACK,
            variables: { id: filter },
          });

          // write back to the cache, updating the data
          cache.writeQuery({
            query: GET_FLASHCARDS_BY_PACK,
            variables: { id: filter },
            data: { 
              fetchFlashcardsByPack: [...fetchFlashcardsByPack, data.createFlashcard.card]
            },
          });
        }}
      >
        {(addFlashcard) => (
          <MediumButton
            className="PackCarousel__button-save"
            disabled={questionText && questionAnswer ? false : true}
            onClick={e => {
              addFlashcard({ variables: { input: {
                text: questionText,
                answer: questionAnswer,
                user_id: "1",
                pack_id: filter,
              }}});
              saveCardAndRefresh();
            }}
            >
              SAVE CARD
            </MediumButton>
        )}
      </Mutation>
    )
  };

  return (
    <PackCarouselWrapper className="PackCarousel">

      <Header className="PackCarousel__header">
        <ButtonGroup className="btn-lbl-combo">
          <BackIcon
            className="PackCarousel__button-back btn-lbl-combo__icon"
            data-testid="button-back"
            onClick={e => exitToPackHome()}
          />
          <Caption3 className="PackHome__button-caption btn-lbl-combo__text">Back</Caption3>
        </ButtonGroup>
      </Header>

      <Counter className="PackCarousel__counter">
        <CounterBody className="counter-container">
          <Subhead className="counter-content">
            {`${index + 1} of ${cards.length}`}
          </Subhead>
        </CounterBody>
      </Counter>

      <CardViewer className="PackCarousel__card-viewer">
        {isReviewMode
          ? (
              <Flashcard
                card={currentQuestion}
              />
          ) : (
              <ComposeFlashcard
                handleTextChange={handleTextChange}
                handleAnswerChange={handleAnswerChange}
                questionText={questionText}
                questionAnswer={questionAnswer}
              />
          )
        }
      </CardViewer>

      <CardNavigation className="PackCarousel__nav">
        {!isReviewMode
          ? (
            <CreateFlashcard />
          ) : (
            <>
              <BackIcon 
                className={`PackCarousel__button-nav${index === 0 ? '--disabled' : ''} back`}
                onClick={null}
                data-testid="button-back"
              />
              <BackIcon 
                className={`PackCarousel__button-nav${index === cards.length - 1 ? '--disabled' : ''} forward`}
                onClick={null}
                data-testid="button-forward"
              />
            </>
          )
        }
      </CardNavigation>

    </PackCarouselWrapper>
  );
};

const PackCarouselWrapper = styled.div`
  display: grid;
  grid-template-rows: repeat(3, auto) minmax(0, 1fr);
  height: 100%;
  background: orange;
`;

const Header = styled.header`
  grid-row: 1 / span 1;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1.5rem 1.5rem 1rem 1.5rem;
  color: ${props => props.theme.font.primary};
`;

const ButtonGroup = styled.div`
  display: flex;
  justify-content: flex-start;
  align-items: center;

  .btn-lbl-combo__icon {
    height: 2rem;
    width: 2rem;
    margin-right: .50rem;

    path {
      fill: ${props => props.theme.color.fonts.eerieBlack};
    }
  }

  .btn-lbl-combo__text {
    color: ${props => props.theme.color.main.primary};
  }
`;

const Counter = styled.section`
  grid-row: 3 / span 1;
  padding: 0rem 1.5rem 0rem 1.5rem;
`;

const CounterBody = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100%;
  padding: .50rem;
  border-top-left-radius: 2rem;
  border-top-right-radius: 2rem;
  background: black;
  color: white;
`;

const CardNavigation = styled.footer`
  grid-row: 2 / span 1;
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  padding: 1rem;

  svg[class~="PackCarousel__button-nav"] {
    height: 2rem;
    path {
      fill: ${props => props.theme.background.special};
    }

    &:hover {
      path {
        fill: ${props => props.theme.background.special};
      }
    }
  }

  svg[class*="--disabled"] {
      path {
        fill: ${props => props.theme.button.default.greyed};
      }
      pointer-events: none;
  }
`;

const BackIcon = styled(Back)`
  height: 2rem;

  &:hover {
    path {
      fill: ${props => props.theme.button.default.alert};
    }
  }
`;

const CardViewer = styled.section`
  grid-row: 4 / span 1;
  padding-bottom: 1rem;
`;


PackCarousel.propTypes = {
  mode: PropTypes.string.isRequired,
  filter: PropTypes.string.isRequired,
  cards: PropTypes.arrayOf(
    PropTypes.object,
  ).isRequired,
};

export default PackCarousel;
