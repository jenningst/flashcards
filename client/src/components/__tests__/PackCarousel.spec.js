import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { PackProvider } from '../../contexts/packContext';
import { MockedProvider } from 'react-apollo/test-utils';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';
import { seedMockPack } from '../../utilities/test-utils';
import { mockCards } from '../../utilities/__mocks/mock-cards';
import { zeroPad } from '../../utilities/helpers';

import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';

import PackCarousel from '../PackCarousel';

afterEach(cleanup);

const renderComponent = ({ theme, mode, filter, cards }) => 
  render(
    <ThemeProvider theme={theme}>
      <PackProvider>
        <MockedProvider addTypename={false}>
          <MemoryRouter initialEntries={['/']}>
            <PackCarousel mode={mode} filter={filter} cards={cards} />
          </MemoryRouter>
        </MockedProvider>
      </PackProvider>
    </ThemeProvider>
  );

describe('<PackCarousel /> spec', () => {
  it.skip('assert component matches snaptshot', () => {
    const mockPack = seedMockPack(mockCards, 1);
    const { asFragment } = renderComponent({ 
      theme: lightTheme,
      mode: mockPack.mode,
      filter: mockPack.filter,
      cards: mockPack.cards, 
    });
    expect(asFragment()).toMatchSnapshot();
  });

  describe('common functionality', () => {
    let mockPack;
    beforeEach(() => {
      mockPack = seedMockPack(mockCards, mockCards.length); 
    });

    it('assert a close button is rendered enabled', () => {
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockPack.mode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByTestId('button-close')).toBeInTheDocument();
    });
  
    it('assert the current question index is rendered at 001', () => {
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockPack.mode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByTestId('current')).toBeInTheDocument();
      expect(getByTestId('current')).toHaveTextContent('001');
    });

    it('assert a counter is rendered initially at the first index', () => {
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockPack.mode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      const totalQuestions = zeroPad(mockPack.cards.length);
      expect(getByTestId('total')).toBeInTheDocument();
      expect(getByTestId('total')).toHaveTextContent(`\ ${totalQuestions}`);
    });

    it('assert navigation buttons are rendered', () => {
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockPack.mode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByTestId('button-back')).toBeInTheDocument();
      expect(getByTestId('button-forward')).toBeInTheDocument();
    });

    it.skip('assert back button is disabled on first card', () => {
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockPack.mode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByTestId('button-back')).toHaveClass('pointer-events', 'none');
    });

    it.todo('assert both buttons enabled when not on first or last card');
    it.todo('assert forward button is disabled on last card');

    it('assert the review mode is displayed when mode is REVIEW MODE', () => {
      const mockMode = 'REVIEW_MODE';
      const prettyMockMode = mockMode.replace('_', ' ');
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockMode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByTestId('mode')).toHaveTextContent(prettyMockMode);
    });

    it('assert the write mode is displayed when mode is WRITE MODE', () => {
      const mockMode = 'WRITE_MODE';
      const prettyMockMode = mockMode.replace('_', ' ');
      const { getByTestId } = renderComponent({ 
        theme: lightTheme,
        mode: mockMode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByTestId('mode')).toHaveTextContent(prettyMockMode);
    });

    it('assert save button is displayed disabled when mode is WRITE MODE', () => {
      const mockMode = 'WRITE_MODE';
      const { getByText } = renderComponent({ 
        theme: lightTheme,
        mode: mockMode,
        filter: mockPack.filter,
        cards: mockPack.cards, 
      });
      expect(getByText(/^save card$/i)).toBeInTheDocument();
      expect(getByText(/^save card$/i)).toBeDisabled();
    });
  });
});