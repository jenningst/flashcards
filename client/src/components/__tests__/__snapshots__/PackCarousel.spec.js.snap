// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackCarousel /> spec assert component matches snaptshot 1`] = `
<DocumentFragment>
  <div
    class="PackCarousel sc-chPdSV giOJoI"
  >
    <header
      class="PackCarousel__header sc-kgoBCf fgeGUY"
    >
      <svg
        class="PackCarousel__button-close sc-kpOJdX guAKNK"
      >
        error.svg
      </svg>
      <div
        class="PackCarousel__mode"
      >
        <p
          class="sc-bZQynM kpviFl"
        >
          REVIEW MODE
        </p>
      </div>
      <div
        class="PackCarousel__counter counter-group sc-ckVGcZ cCaBEt"
      >
        <div
          class="counter-group__current"
        >
          <h1
            class="sc-ifAKCX fUpSho"
          >
            001
          </h1>
        </div>
        <div
          class="counter-group__total"
        >
          <p
            class="sc-bZQynM kpviFl"
          >
            / 001
          </p>
        </div>
        <div
          class="counter-group__indicator"
        >
          <svg
            class="active-dot"
          >
            oval.svg
          </svg>
        </div>
      </div>
    </header>
    <section
      class="PackCarousel__card-viewer sc-dxgOiQ gVmsQE"
    >
      <div
        class="Flashcard sc-VigVT yrdFj"
      >
        <section
          class="Flashcard__content ContentArea sc-jTzLTM kljkwS"
        >
          <p
            class="ContentArea__text sc-EHOje gzeaHE"
            data-testid="card-text"
          >
            Did you hear about the guy who lost the left side of his body
          </p>
        </section>
        <button
          class="sc-gZMcBi Flashcard__button-toggle sc-fjdhpX jHoKHV"
          data-testid="card-button"
          type="button"
        >
          SHOW ANSWER
        </button>
      </div>
    </section>
    <footer
      class="PackCarousel__nav sc-kGXeez hGClzM"
    >
      <svg
        class="PackCarousel__button-nav disabled back"
      >
        left-arrow.svg
      </svg>
      <svg
        class="PackCarousel__button-nav disabled forward"
      >
        right-arrow.svg
      </svg>
    </footer>
  </div>
</DocumentFragment>
`;
